sssd_test_framework/hosts/keycloak.py

```diff
@@ -0,0 +1,108 @@
+"""Keycloak multihost host."""
+
+from __future__ import annotations
+
+import time
+
+from pytest_mh.ssh import SSHProcessError
+
+from .base import BaseDomainHost
+
+__all__ = [
+    "KeycloakHost",
+]
+
+
+class KeycloakHost(BaseDomainHost):
+    """
+    Keycloak host object.
+
+    Provides features specific for Keycloak server.
+
+    This class adds ``config.adminpw`` multihost configuration option to set
+    password of the Keycloak admin user so we run kcadm.sh commands to set
+    options in Keycloak.
+
+    .. code-block:: yaml
+        :caption: Example multihost configuration
+        :emphasize-lines: 4
+
+        - hostname: master.keycloak.test
+          role: keycloak
+          config:
+            adminpw: Secret123
+
+    .. note::
+
+        Full backup and restore is supported. However, the operation relies on
+        ``kc.sh export`` and ``kc.sh import`` commands which can take several
+        seconds to finish.
+    """
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+
+        self.adminpw = self.config.get("adminpw", "Secret123")
+
+        # Backup of original data
+        self.__backup: str | None = None
+
+    def kclogin(self) -> None:
+        """
+        Obtain ``admin`` user credentials for Keycloak
+        """
+        for x in range(0, 5):
+            try:
+                str_error = None
+                self.ssh.exec(
+                    [
+                        "/opt/keycloak/bin/kcadm.sh",
+                        "config",
+                        "credentials",
+                        "--server",
+                        f"https://{self.hostname}:8443/auth/",
+                        "--realm",
+                        "master",
+                        "--user",
+                        "admin",
+                        "--password",
+                        self.adminpw,
+                    ]
+                )
+            except SSHProcessError as err:
+                str_error = str(err)
+
+            if str_error:
+                time.sleep(30)
+            else:
+                break
+
+    def backup(self) -> None:
+        """
+        Backup all Keycloak server data.
+
+        This is done by calling ``kc.sh export`` on the server
+        and can take several seconds to finish.
+        """
+        if self.__backup is not None:
+            return
+
+        self.ssh.run("systemctl stop keycloak")
+        self.ssh.run("/opt/keycloak/bin/kc.sh export --dir /tmp/kcbackup")
+        self.ssh.run("systemctl start keycloak")
+        cmd = self.ssh.run("ls -1 /tmp/kcbackup | tail -n 1")
```

Comment:
It might be better to put it into single ssh call to safe resources.

Reply:
Updating but, should I leave the `ls`  separate to avoid potential issues with output from a previous command affecting the `ls` used to determine _backup path?

Reply:
Is there any output on successful run? If yes, then sure. Otherwise `set -e` should suffice. But I'll leave it up to you.

Reply:
There is output on successful run from the export command but I think I resolved that with a redirect (which may be helpful debugging later too).